File: main_XYplots_and_DPEstMCMC.m
% This is the script that generates Figures 4, 5, 6 and Table 1 in the
% paper "MCMC for Bayesian estimation of Differential Privacy from
% Membership Inference Attacks"

% Read the file
clear all; clc; close all; fc = 0;
rng(1);

%% Setup and initizalization
foldername = 'Output_Figures';
if ~exist(foldername, 'dir')
    mkdir(foldername);
end

% Files where the loss metrics are stored
filenames = {
    'dp_attack_wo_DP_fixed_weights_0.1.txt', ...
    'dp_attack_wo_DP_random_weights_0.1.txt', ...
    'dp_attack_w_DP_fixed_weights_0.1.txt', ...
    'dp_attack_w_DP_random_weights_0.1.txt', ...
    'dp_attack_w_DP_random_weights_0.05.txt', ...
    'dp_attack_w_DP_random_weights_0.01.txt'
    };

alg_names = {'$\mathcal{A}_{1}$', '$\mathcal{A}_{2}$',...
    '$\mathcal{A}_{3}$, $\sigma = 0.1$', ...
    '$\mathcal{A}_{4}$, $\sigma = 0.1$',...
    '$\mathcal{A}_{4}$, $\sigma = 0.05$', ...
    '$\mathcal{A}_{4}$, $\sigma = 0.01$'};

L_f = length(filenames);

%% MCMC parameters
M = 100000;  % MCMC iterations
K = 1000; % Number of auxiliary variables
delta_DP = 0.001;

sigma_q_vec = [0.1, 0.01, 0.001, 0.001];

% hyperparameters: [eps_var, sab, tau_var, rho_var];
theta_hyper = [10, 1, 0.0001, 0.01];

% initial value for [eps_DP, s, tau, rho]
theta = [1, 0.5, 0, 0];

update_params = [1, 1, 1, 1];

% target alpha values
alpha_vec = 0.01:0.01:0.99;
L_a = length(alpha_vec);

% 20 MIAs each tested 100 times with H0 and 100 times with H1
n = 20; N = 100;

% Initialization
eps_DP_low = zeros(1, L_f);
eps_DP_high = zeros(1, L_f);
s_low = zeros(1, L_f);
s_high = zeros(1, L_f);

X = repmat({repmat({zeros(N, L_a)}, 1, n)}, 1, L_f);
Y = repmat({repmat({zeros(N, L_a)}, 1, n)}, 1, L_f);
E = repmat({repmat({zeros(N, L_a)}, 1, n)}, 1, L_f);


Eps_DP_MCMC = cell(1, L_f);
S_MCMC = cell(1, L_f);
Tau_MCMC = cell(1, L_f);
Rho_MCMC = cell(1, L_f);

acf = cell(1, L_f);
lags = cell(1, L_f);

for fn = 1:L_f
    filename = filenames{fn};

    [l_0_matrix, l_1_matrix] = read_losses(filename);

    fprintf('Calculating errors for file fn: %d \n', fn);
    % Find x's
    for h = [0, 1]
        for k = 1:n % challenge base
            for j = 1:N % challange
                if h == 0
                    % This point is challenged
                    l_ast = l_0_matrix(k, j);
                    % The rest are used to learn H0 and H1
                    l0 = l_0_matrix(k, [1:(j-1), (j+1):end]);
                    l1 = l_1_matrix(k, :);
                else
                    % This point is challenged
                    l_ast = l_1_matrix(k, j);
                    % The rest are used to learn H0 and H1
                    l1 = l_1_matrix(k, [1:(j-1), (j+1):end]);
                    l0 = l_0_matrix(k, :);
                end
                % Learn H0 and H1
                mu_l0 = mean(l0); mu_l1 = mean(l1);
                var_ell0 = var(l0); var_ell1 = var(l1);

                % Decisions of LRT with the fitted H0 and H1
                D = LRT_normal(l_ast, mu_l0, var_ell0, mu_l1, var_ell1, alpha_vec);

                if h == 0 % H0 is true
                    X{fn}{k}(j, :) = D;
                else % H1 is true
                    Y{fn}{k}(j, :) = 1-D;
                end
            end
        end
    end
    % Total error counts
    for k = 1:n
        E{fn}{k} = X{fn}{k} + Y{fn}{k};
    end

    %% Plot the error lines
    fc = fc + 1;
    f = figure(fc);
    set(gcf, 'Visible', 'off'); % Hide it
    f.Position = [100 100 200 170];
    set(gcf, 'renderer', 'painters');

    for k = 1:n
        plot(mean(X{fn}{k}), mean(Y{fn}{k}), '.-');
        hold on;
    end
    hold off;
    grid on;

    exportgraphics(gcf,[foldername, sprintf('/Errors_%s.pdf', filename)]);
    close(f);

    %% Run the MCMC algorithm to estimate epsilon and s
    X0f = zeros(1, n);
    Y0f = zeros(1, n);
    N0 = 100*ones(1, n);
    N1 = 100*ones(1, n);

    % Get the (X, Y) values at alpha_t = 0.1
    for k = 1:n
        X0f(k) = sum(X{fn}{k}(:, 10));
        Y0f(k) = sum(Y{fn}{k}(:, 10));
    end

    fprintf('Running MCMC for fn: %d \n', fn);

    tic;
    [theta_samps] = MCMC_epsDP_v2(N0, N1, X0f, Y0f, M, K, delta_DP, ...
        theta, sigma_q_vec, theta_hyper, update_params, 'bivariate');

    eps_DP_samps = theta_samps(1, :);
    s_samps = theta_samps(2, :);
    tau_samps = theta_samps(3, :);
    rho_samps = theta_samps(4, :);

    toc;
    % Store the estimates epsilon and s
    Eps_DP_MCMC{fn} = eps_DP_samps;
    S_MCMC{fn} = s_samps;
    Tau_MCMC{fn} = tau_samps;
    Rho_MCMC{fn} = rho_samps;

    eps_DP_after_burn_in = eps_DP_samps(M/5:M);
    
    [acf{fn}, lags{fn}] = autocorr(eps_DP_after_burn_in, 'Numlags', min(M-1, 5000));

    % Get the credible intervals for epsilon and s
    eps_DP_low(fn) = quantile(eps_DP_after_burn_in, 0.05);
    eps_DP_high(fn) = quantile(eps_DP_after_burn_in, 0.95);

    s_after_burn_in = s_samps(M/10:M);
    s_low(fn) = quantile(s_after_burn_in, 0.05);
    s_high(fn) = quantile(s_after_burn_in, 0.95);

end

%%
fc = fc + 1;
f = figure(fc);
set(gcf, 'Visible', 'off'); % Hide it
f.Position = [100 100 200 100];
for fn = 1:L_f
    set(gcf, 'renderer', 'painters');
    plot(lags{fn}, acf{fn});
    hold on;
end
xlabel('lag'); ylabel('Sample ACF');
set(gca, 'xlim', [0, 5000]);

grid on;
exportgraphics(gcf,[foldername '/auto_corr.pdf']);
close(f);

save('real_data_experiments');



%% plot the samples

f = figure(fc);
f.Position = [100 100 800 500];

for fn = 1:L_f

    subplot(6, 4, (fn-1)*4 + 1);
    plot(Eps_DP_MCMC{fn});
    ylabel(alg_names{fn}, 'Interpreter', 'Latex');
    if fn < L_f
        set(gca, 'xticklabel', []);
    else
        xlabel('iteration');
    end
    title('$\epsilon$', 'Interpreter', 'Latex');
    subplot(6, 4, (fn-1)*4 + 2);
    plot(S_MCMC{fn});
    title('$s$', 'Interpreter', 'Latex');
    if fn < L_f
        set(gca, 'xticklabel', []);
    else
        xlabel('iteration');
    end
    subplot(6, 4, (fn-1)*4 + 3);
    plot(Tau_MCMC{fn});
    title('$\tau$', 'Interpreter', 'Latex');
    if fn < L_f
        set(gca, 'xticklabel', []);
    else
        xlabel('iteration');
    end
    subplot(6, 4, (fn-1)*4 + 4);
    plot(Rho_MCMC{fn});
    title('$\rho$', 'Interpreter', 'Latex');
    
    if fn < L_f
        set(gca, 'xticklabel', []);
    else
        xlabel('iteration');
    end
end
    
exportgraphics(gcf,[foldername '/MCMC_samples.pdf']);

%% Draw the 2D histograms
for fn = 1:L_f
 % Plot and save
    filename = filenames{fn};
    fc = fc + 1;
    f = figure(fc);
    set(gcf, 'Visible', 'on'); % Hide it
    f.Position = [100 100 200 170];
    set(gcf, 'renderer', 'painters');
    colormap(gray);
    histogram2(Eps_DP_MCMC{fn}(end/2:end), S_MCMC{fn}(end/2:end), 'DisplayStyle','tile','ShowEmptyBins','on');
    xlabel('$\epsilon$', 'Interpreter', 'Latex', 'Fontsize', 14);
    ylabel('$s$', 'Interpreter', 'Latex', 'Fontsize', 14);

    exportgraphics(gcf,sprintf('%s/histogram_%s.pdf',foldername, filename));
    close(f);
end

File: python/experiments/main_XYplots_and_DPEst_MCMC.py
# main_XYplots_and_DPEstMCMC.py
import numpy as np
import matplotlib.pyplot as plt
import os
import time
from statsmodels.tsa.stattools import acf

# Import translated functions
from src.read_losses import read_losses
from src.LRT_normal import LRT_normal
from src.MCMC_epsDP_v2 import MCMC_epsDP_v2

def run_real_data_experiments():
    """
    This script generates Figures 4, 5, 6 and Table 1 from the paper.
    It requires the specified data files to be in the same directory.
    """
    print("Starting real data experiments...")
    # Set up matplotlib for LaTeX rendering
    plt.rcParams.update({
        "text.usetex": True,
        "font.family": "serif",
        "font.serif": ["Computer Modern Roman"],
        "font.size": 10
    })

    # --- Setup and Initialization ---
    np.random.seed(1)
    foldername = 'Output_Figures_Python'
    os.makedirs(foldername, exist_ok=True)

    # Files where the loss metrics are stored
    filenames = [
        'dp_attack_wo_DP_fixed_weights_0.1.txt',
        'dp_attack_wo_DP_random_weights_0.1.txt',
        'dp_attack_w_DP_fixed_weights_0.1.txt',
        'dp_attack_w_DP_random_weights_0.1.txt',
        'dp_attack_w_DP_random_weights_0.05.txt',
        'dp_attack_w_DP_random_weights_0.01.txt'
    ]
    # Check if data files exist before proceeding
    for f in filenames:
        if not os.path.exists(f'../data/{f}'):
            print(f"Error: Data file '{f}' not found. This script cannot run without it.")
            print("Please place the required data files in the same directory.")
            return

    alg_names = [
        r'$\mathcal{A}_{1}$', r'$\mathcal{A}_{2}$',
        r'$\mathcal{A}_{3}$, $\sigma = 0.1$',
        r'$\mathcal{A}_{4}$, $\sigma = 0.1$',
        r'$\mathcal{A}_{4}$, $\sigma = 0.05$',
        r'$\mathcal{A}_{4}$, $\sigma = 0.01$'
    ]
    L_f = len(filenames)

    # MCMC parameters
    M = 100_000  # MCMC iterations
    K = 1000     # Number of auxiliary variables
    delta_DP = 0.001
    sigma_q_vec = [0.1, 0.01, 0.001, 0.001]
    theta_hyper = [10, 1, 0.0001, 0.01]  # [eps_var, sab, tau_var, rho_var]
    theta_init = [1, 0.5, 0, 0] # [eps_DP, s, tau, rho]
    update_params = [1, 1, 1, 1]

    # Target alpha values
    alpha_vec = np.arange(0.01, 1.0, 0.01)
    L_a = len(alpha_vec)
    n, N = 20, 100 # 20 MIAs each tested 100 times with H0 and 100 times with H1

    # Initialization
    # Using lists of lists of numpy arrays, equivalent to MATLAB's nested cell arrays
    X = [[np.zeros((N, L_a)) for _ in range(n)] for _ in range(L_f)]
    Y = [[np.zeros((N, L_a)) for _ in range(n)] for _ in range(L_f)]
    E = [[np.zeros((N, L_a)) for _ in range(n)] for _ in range(L_f)]

    Eps_DP_MCMC, S_MCMC, Tau_MCMC, Rho_MCMC = [[] for _ in range(4)]
    acf_results, lags_results = [], []

    for fn, filename in enumerate(filenames):
        print(f"\n--- Processing file: {filename} ({fn+1}/{L_f}) ---")
        print(f'../data/{filename}')
        l_0_matrix, l_1_matrix = read_losses(f'../data/{filename}')

        print("Calculating errors (Type I/II)...")
        for h in [0, 1]:  # 0 for H0 is true, 1 for H1 is true
            for k in range(n):  # Challenge base
                for j in range(N):  # Challenge instance
                    if h == 0: # H0 is true
                        l_ast = l_0_matrix[k, j]
                        l0 = np.concatenate((l_0_matrix[k, :j], l_0_matrix[k, j+1:]))
                        l1 = l_1_matrix[k, :]
                    else: # H1 is true
                        l_ast = l_1_matrix[k, j]
                        l1 = np.concatenate((l_1_matrix[k, :j], l_1_matrix[k, j+1:]))
                        l0 = l_0_matrix[k, :]
            
                    mu_l0, var_l0 = np.mean(l0), np.var(l0, ddof=1)
                    mu_l1, var_l1 = np.mean(l1), np.var(l1, ddof=1)
                    
                    D = LRT_normal(l_ast, mu_l0, var_l0, mu_l1, var_l1, alpha_vec)

                    if h == 0:
                        X[fn][k][j, :] = D  # False Positive Rate (alpha)
                    else:
                        Y[fn][k][j, :] = 1 - D # False Negative Rate (beta)
        
        for k in range(n):
            E[fn][k] = X[fn][k] + Y[fn][k]

        # Plot the error lines (alpha vs beta)
        plt.figure(figsize=(5, 4.5))
        for k in range(n):
            plt.plot(np.mean(X[fn][k], axis=0), np.mean(Y[fn][k], axis=0), '.-')
        plt.grid(True)
        plt.xlabel('$\\alpha$ (False Positive Rate)')
        plt.ylabel('$\\beta$ (False Negative Rate)')
        plt.title(f'Attack Performance for {alg_names[fn]}')
        plt.savefig(os.path.join(foldername, f'Errors_{filename}.pdf'))
        plt.close()

        # Get the (X, Y) error counts at alpha_target = 0.1
        alpha_target_idx = 9 # Corresponds to alpha = 0.1
        X0f = np.array([np.sum(X[fn][k][:, alpha_target_idx]) for k in range(n)])
        Y0f = np.array([np.sum(Y[fn][k][:, alpha_target_idx]) for k in range(n)])
        N0 = 100 * np.ones(n)
        N1 = 100 * np.ones(n)

        print(f"Running MCMC for file: {filename}...")
        start_time = time.time()
        theta_samps = MCMC_epsDP_v2(N0, N1, X0f, Y0f, M, K, delta_DP,
                                    theta_init, sigma_q_vec, theta_hyper, update_params, 'bivariate')
        print(f"MCMC finished in {time.time() - start_time:.2f} seconds.")

        Eps_DP_MCMC.append(theta_samps[0, :])
        S_MCMC.append(theta_samps[1, :])
        Tau_MCMC.append(theta_samps[2, :])
        Rho_MCMC.append(theta_samps[3, :])

        # Autocorrelation
        eps_after_burn_in = theta_samps[0, M//5:]
        acf_vals, confint = acf(eps_after_burn_in, nlags=min(len(eps_after_burn_in)-1, 5000), alpha=0.05, fft=True)
        lags = np.arange(len(acf_vals))
        acf_results.append(acf_vals)
        lags_results.append(lags)
        
        # 2D Histogram
        plt.figure(figsize=(5, 4.5))
        plt.hist2d(theta_samps[0, M//2:], theta_samps[1, M//2:], bins=50, cmap='gray', density=True)
        plt.xlabel('$\\epsilon$')
        plt.ylabel('$s$')
        plt.title(f'Posterior for {alg_names[fn]}')
        plt.colorbar(label='Frequency')
        plt.savefig(os.path.join(foldername, f'histogram_{filename}.pdf'))
        plt.close()

    # --- Final Aggregate Plots ---
    # Autocorrelation Plot
    plt.figure(figsize=(6, 4))
    for fn in range(L_f):
        plt.plot(lags_results[fn], acf_results[fn], label=alg_names[fn])
    plt.xlabel('Lag')
    plt.ylabel('Sample ACF')
    plt.xlim(0, 5000)
    plt.grid(True)
    plt.legend(fontsize='small')
    plt.savefig(os.path.join(foldername, 'auto_corr.pdf'))
    plt.close()
    
    # MCMC trace plots
    fig, axes = plt.subplots(L_f, 4, figsize=(15, 12), sharex=True)
    param_titles = ['$\\epsilon$', '$s$', '$\\tau$', '$\\rho$']
    for fn in range(L_f):
        axes[fn, 0].plot(Eps_DP_MCMC[fn])
        axes[fn, 0].set_ylabel(alg_names[fn], rotation=0, labelpad=40, va='center')
        
        axes[fn, 1].plot(S_MCMC[fn])
        axes[fn, 2].plot(Tau_MCMC[fn])
        axes[fn, 3].plot(Rho_MCMC[fn])
        
        if fn == 0:
            for j in range(4):
                axes[fn, j].set_title(param_titles[j])
        if fn == L_f - 1:
            for j in range(4):
                axes[fn, j].set_xlabel('Iteration')
    
    plt.tight_layout(rect=[0.05, 0, 1, 1])
    plt.savefig(os.path.join(foldername, 'MCMC_samples.pdf'))
    plt.close()

    # Save results to a file
    np.savez(os.path.join(foldername, 'real_data_experiments.npz'),
             Eps_DP_MCMC=np.array(Eps_DP_MCMC, dtype=object),
             S_MCMC=np.array(S_MCMC, dtype=object),
             Tau_MCMC=np.array(Tau_MCMC, dtype=object),
             Rho_MCMC=np.array(Rho_MCMC, dtype=object),
             acf=np.array(acf_results, dtype=object),
             lags=np.array(lags_results, dtype=object))
    
    print(f"\nAll experiments finished. Outputs are saved in '{foldername}' directory.")

if __name__ == '__main__':
    run_real_data_experiments()